codalab/model/worker_model.py

```diff
@@ -0,0 +1,322 @@
+        for row in worker_dependency_rows:
+            worker_dict[(row.user_id, row.worker_id)]['dependency_uuids'].append(row.dependency_uuid)
+        return worker_dict.values()
+
+    def get_bundle_worker(self, uuid):
+        """
+        Returns information about the worker that the given bundle is running
+        on. This method should be called only for bundles that are running.
+        """
+        with self.engine.begin() as connection:
+            row = connection.execute(cl_worker_run.select()
+                                     .where(cl_worker_run.c.run_uuid == uuid)).fetchone()
+            precondition(row, 'Trying to find worker for bundle that is not running.')
+            if row:
+                return {
+                    'user_id': row.user_id,
+                    'worker_id': row.worker_id,
+                    'socket_id': row.socket_id,
+                }
+
+    def allocate_socket(self, user_id, worker_id, conn=None):
+        """
+        Allocates a unique socket ID.
+        """
+        def do(conn):
+            socket_row = {
+               'user_id': user_id,
+               'worker_id': worker_id,
+            }
+            return conn.execute(
+                cl_worker_socket.insert().values(socket_row)
+            ).inserted_primary_key[0]
+        if conn is None:
+            with self._engine.begin() as conn:
+                return do(conn)
+        else:
+            return do(conn)
+
+    def deallocate_socket(self, socket_id):
+        """
+        Cleans up the socket, removing the associated file in the socket
+        directory.
+        """
+        self._cleanup_socket(socket_id)
+        with self._engine.begin() as conn:
+            conn.execute(cl_worker_socket.delete()
+                         .where(cl_worker_socket.c.socket_id == socket_id))
+
+    def _socket_path(self, socket_id):
+        return os.path.join(self._socket_dir, str(socket_id))
+
+    def _cleanup_socket(self, socket_id):
+        try:
+            os.remove(self._socket_path(socket_id))
+        except OSError:
+            pass
+
+    def start_listening(self, socket_id):
+        """
+        Returns a Python socket object that can be used to accept connections.
+        
+        This method should be called externally only by code that needs to
+        receive more than 1 message on the same socket. The returned object
+        should be passed as the sock parameter to one of the get methods.
+
+        For code that needs to receive only a single message, use the get
+        methods without providing a socket object.
+        """
+        self._cleanup_socket(socket_id)
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.bind(self._socket_path(socket_id))
+        sock.listen(0)
+        return sock
+
+    def get_stream(self, socket_id, timeout_secs, sock=None):
+        """
+        Receives a single message on the socket with the given ID and returns a
```

Comment:
Why can't we just have this method always just take a sock, and it's up to the caller to call start_listening on socket_id?